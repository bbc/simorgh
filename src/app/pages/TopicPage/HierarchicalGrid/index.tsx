/** @jsx jsx */

import { css, jsx, Theme } from '@emotion/react';
import Promo from '../../../legacy/components/Promo';
import { DESKTOP, TABLET, MOBILE, SMALL } from './dataStructures';
import { styles } from './index.styles';

type Summary = {
  title: string;
  description?: string;
  type: string;
  id: string;
  link?: string;
  firstPublished?: string | number;
  mediaDuration?: string | number;
  imageUrl?: string;
  imageAlt?: string;
  mediaType?: 'audio' | 'video' | 'photogallery';
};

type Summaries = {
  summaries: Summary[];
};

const getStyles = (promoCount: number, i: number, mq: any) => {
  return css({
    [mq.GROUP_1_MAX_WIDTH]: {
      ...SMALL[promoCount - 1][i],
    },
    [mq.GROUP_2_ONLY]: {
      ...MOBILE[promoCount - 1][i],
    },
    [mq.GROUP_3_ONLY]: {
      ...TABLET[promoCount - 1][i],
    },
    [mq.GROUP_4_MIN_WIDTH]: {
      ...DESKTOP[promoCount - 1][i],
    },
  });
};

const HiearchicalGrid = ({ summaries }: Summaries) => {
  if (!summaries || summaries.length < 3) return null;
  const summaryItems = summaries.slice(0, 12);
  return (
    <ul role="list" css={styles.list}>
      {summaryItems.map((promo, i) => {
        return (
          <li
            key={promo.id}
            css={({ mq }: Theme) => [
              styles.item,
              getStyles(summaryItems.length, i, mq),
            ]}
          >
            <Promo>
              <Promo.Image
                src={promo.imageUrl || ''}
                alt={promo.imageAlt}
                loading="lazy"
              >
                <Promo.MediaIcon type={promo.mediaType}>
                  {promo.mediaDuration}
                </Promo.MediaIcon>
              </Promo.Image>
              <Promo.Heading
                css={(theme: Theme) => i === 0 && theme.fontSizes.paragon}
              >
                <Promo.A href={promo.link}>{promo.title}</Promo.A>
              </Promo.Heading>
              <Promo.Body className="promo-paragraph" css={styles.body}>
                {promo.description}
              </Promo.Body>
              <Promo.Timestamp className="promo-timestamp">
                {promo.firstPublished}
              </Promo.Timestamp>
            </Promo>
          </li>
        );
      })}
    </ul>
  );
};

export default HiearchicalGrid;
