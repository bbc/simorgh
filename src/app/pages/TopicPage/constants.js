export const VISUAL_STYLE = {
  NONE: 'NONE',
  BANNER: 'BANNER',
  COLLECTION: 'COLLECTION',
  LINKS: 'LINKS',
  FEED: 'FEED',
  RANKED: 'RANKED',
};

export const VISUAL_PROMINENCE = {
  MINIMUM: 'MINIMUM',
  LOW: 'LOW',
  NORMAL: 'NORMAL',
  HIGH: 'HIGH',
  MAXIMUM: 'MAXIMUM',
};

export const COMPONENT_NAMES = {
  MESSAGE_BANNER: 'message-banner',
  SIMPLE_CURATION_GRID: 'simple-curation-grid',
  HIERARCHICAL_CURATION_GRID: 'hierarchical-curation-grid',
  NOT_SUPPORTED: 'not-supported',
};
