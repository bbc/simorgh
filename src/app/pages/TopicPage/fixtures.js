export const amharicSingleItem = {
  title: 'ዶናልድ ትራምፕ',
  description: 'ዶናልድ ትራምፕ',
  curations: [
    {
      summaries: [
        {
          title: 'በጥቅም ላይ የዋሉ የጥንታዊ ግብፅ የህክምና ጥበባት',
          type: 'article',
          firstPublished: '2017-08-22T08:41:04.000Z',
          link: 'https://www.bbc.com/amharic/news-23143874',
          imageUrl: '',
          imageAlt: '',
          id: 'c31dd8c5-2e55-9143-8ed2-5f9c15d6c91c',
        },
      ],
      activePage: 1,
      pageCount: 1,
      curationId: 'urn:bbc:vivo:curation:bf45c519-cfee-4627-832f-41759cdeb30c',
      curationType: 'vivo-stream',
      visualProminence: 'NORMAL',
    },
  ],
  metadata: { type: 'Topic' },
};

export const pidginMultipleItems = {
  title: 'Donald Trump',
  description: 'Donald Trump',
  imageData: null,
  curations: [
    {
      summaries: [
        {
          title: 'Trump social-media platform don go live',
          type: 'article',
          firstPublished: '2022-02-21T15:32:39.000Z',
          link: 'https://www.bbc.com/pidgin/world-60465144',
          imageUrl:
            'https://ichef.bbci.co.uk/news/{width}/cpsprodpb/2C6A/production/_123307311_trump__truth_index_976-nc.png',
          imageAlt:
            "Truth Social will be run by the former US president's newly created Trump Media & Technology Group",
          id: '92422e0f-00eb-47f9-9795-696d826313b1',
        },
        {
          title: 'Wetin happun for January 6 one year ago?',
          type: 'article',
          firstPublished: '2022-01-06T19:00:29.000Z',
          link: 'https://www.bbc.com/pidgin/tori-59901959',
          imageUrl:
            'https://ichef.bbci.co.uk/news/{width}/cpsprodpb/DE3A/production/_122609865_january6timelinewetinhappunforjanuary6oneyearago.jpg',
          imageAlt:
            'January 6 timeline: Wetin happun for January 6 one year ago?',
          id: 'e2263a1c-8d5a-4a73-a00c-881acfa34381',
        },
        {
          title: 'Why Donald Trump say Bitcoin na scam',
          type: 'article',
          firstPublished: '2021-06-08T07:41:51.000Z',
          link: 'https://www.bbc.com/pidgin/tori-57396574',
          imageUrl:
            'https://ichef.bbci.co.uk/news/{width}/cpsprodpb/10BB/production/_118838240_3164b6b1-d415-43c8-9bce-3e497aac61bf.jpg',
          imageAlt: 'Donald Trump',
          id: '11cd8a9d-0d67-4b92-b20b-df71740a4fdd',
        },
        {
          title: "Wetin be dis new 'website' wey Donald Trump launch?",
          type: 'article',
          firstPublished: '2021-05-05T09:12:36.000Z',
          link: 'https://www.bbc.com/pidgin/tori-56991816',
          imageUrl:
            'https://ichef.bbci.co.uk/news/{width}/cpsprodpb/D8D0/production/_116340555_gettyimages-1230086798.jpg',
          imageAlt: 'Trump',
          id: '682b5bcd-0dfd-4e9f-b1a0-14dfe206df83',
        },
      ],
      activePage: 1,
      pageCount: 6,
      curationId: 'urn:bbc:vivo:curation:bf45c519-cfee-4627-812f-41759cdeb30c',
      curationType: 'vivo-stream',
      visualProminence: 'NORMAL',
    },
  ],
  metadata: { type: 'Topic' },
};

export const mundoMultipleCurations = {
  title: 'Colombia Peace Agreement',
  description: 'Peace agreement between the Colombian government and FARC',
  imageData: {
    url: 'https://ichef.bbci.co.uk/news/png/480/cpsdevpb/162f/test/04adc6f0-1312-11ed-adc3-595fb3bc204b.png',
    urlTemplate:
      'https://ichef.bbci.co.uk/news/png/{width}/cpsdevpb/162f/test/04adc6f0-1312-11ed-adc3-595fb3bc204b.png',
    originalSize: { width: 456, height: 456 },
    altText: 'Highland Cow',
    copyright: 'RM',
    type: 'badge',
  },
  curations: [
    {
      summaries: [
        {
          title: 'El poder del "chilango" llegó a Londres 14',
          type: 'article',
          firstPublished: '2016-05-05T07:38:30.000Z',
          link: 'https://www.bbc.com/mundo/vert-earth-23038377',
          imageUrl:
            'https://ichef.test.bbci.co.uk/news/{width}/cpsdevpb/17654/test/_63482859_orange1.jpg',
          description:
            'A Los de Abajo los encontramos en el legendario Marquee de Londres, casa de grandes como Jimi Hendrix, U2, REM y The Who.',
          imageAlt: 'Orange 1',
          id: 'ae7c9f51-3144-ba4b-a881-f88822b36c8c',
        },
        {
          title:
            'El conflictivo pasado con Venezuela de Rex Tillerson, nuevo secretario de Estado de EEUU',
          type: 'article',
          firstPublished: '2017-02-03T11:26:55.000Z',
          link: 'https://www.bbc.com/mundo/23100317',
          imageUrl:
            'https://ichef.test.bbci.co.uk/news/{width}/cpsdevpb/167B2/test/_63528029_036820129-1.jpg',
          description:
            'Es posiblemente en lo único que se ponen de acuerdo gobierno y oposición en Venezuela: ambos tienen esperanzas en una mejor relación con Estados Unidos y que eso les ayude en sus objetivos contrapuestos.',
          imageAlt:
            'ExxonMobil Chairman and CEO Rex Tillerson speaks during the IHS CERAWeek 2015 energy conference in Houston, Texas April 21, 2015',
          id: 'ff4f3677-0f70-574e-b129-6b3396788390',
        },
      ],
      activePage: 1,
      pageCount: 1,
      curationId: 'urn:bbc:tipo:list:d0348c91-39f5-4224-91cf-6ca9f2812376',
      curationType: 'tipo-curation',
      position: 0,
      title: 'Analysis',
      visualProminence: 'NORMAL',
    },
    {
      summaries: [
        {
          title: 'El video de prueba',
          type: 'video',
          firstPublished: '2022-05-12T10:41:21.000Z',
          link: 'https://www.bbc.com/mundo/23048701',
          imageUrl:
            'https://ichef.test.bbci.co.uk/news/{width}/cpsdevpb/FA0D/test/_63931046_p00wjk8v.jpg',
          description: 'Es porque le falta el sumario?',
          imageAlt: 'Al Tag',
          id: 'c696c8a1-8939-3e4b-b87b-5aa1de1b4c96',
        },
        {
          title: 'El video de prueba',
          type: 'video',
          firstPublished: '2022-05-12T10:41:21.000Z',
          link: 'https://www.bbc.com/mundo/23048701',
          imageUrl:
            'https://ichef.test.bbci.co.uk/news/{width}/cpsdevpb/FA0D/test/_63931046_p00wjk8v.jpg',
          description: 'Es porque le falta el sumario?',
          imageAlt: 'Al Tag',
          id: 'c696c8a1-8939-34k2-b87b-5aa1de1b4c96',
        },
      ],
      activePage: 1,
      pageCount: 1,
      curationId: 'urn:bbc:vivo:curation:22f2633b-f31e-4195-81d6-5dd5f766ca42',
      curationType: 'vivo-stream',
      position: 1,
      title: 'Video',
      visualProminence: 'NORMAL',
    },
  ],
  activePage: null,
  pageCount: null,
  variantTopicId: null,
  contentType: 'application/json; charset=utf-8',
};

export const mundoWithBadgeAndDescr = {
  title: 'Colombia Peace Agreement',
  description: 'Peace agreement between the Colombian government and FARC',
  imageData: {
    url: 'https://ichef.bbci.co.uk/news/png/480/cpsdevpb/b40b/test/5e614490-0360-11ed-b35d-c5a474731c9c.png',
    urlTemplate:
      'https://ichef.bbci.co.uk/news/png/{width}/cpsdevpb/b40b/test/5e614490-0360-11ed-b35d-c5a474731c9c.png',
    originalSize: { width: 1200, height: 800 },
    altText: 'Colombia flag',
    copyright: 'Wikipedia',
    type: 'badge',
  },
  curations: [
    {
      summaries: [
        {
          title:
            'Qué hace que los chilenos sean tan resistentes a los terremotos',
          type: 'article',
          firstPublished: '2017-02-03T11:38:04.000Z',
          link: 'https://www.bbc.com/mundo/23100229',
          imageUrl:
            'https://ichef.test.bbci.co.uk/news/{width}/cpsdevpb/3AEA/test/_63528051_000870367-1.jpg',
          imageAlt: 'Santiago de Chile',
          id: '1ff8ff1a-9570-4b44-b90a-73cabca969bb',
        },
        {
          title:
            'El conflictivo pasado con Venezuela de Rex Tillerson, nuevo secretario de Estado de EEUU',
          type: 'article',
          firstPublished: '2017-02-03T11:26:55.000Z',
          link: 'https://www.bbc.com/mundo/23100317',
          imageUrl:
            'https://ichef.test.bbci.co.uk/news/{width}/cpsdevpb/167B2/test/_63528029_036820129-1.jpg',
          imageAlt:
            'ExxonMobil Chairman and CEO Rex Tillerson speaks during the IHS CERAWeek 2015 energy conference in Houston, Texas April 21, 2015',
          id: 'ff4f3677-0f70-574e-b129-6b3396788390',
        },
        {
          title: 'El poder del "chilango" llegó a Londres 14',
          type: 'article',
          firstPublished: '2016-05-05T07:38:30.000Z',
          link: 'https://www.bbc.com/mundo/vert-earth-23038377',
          imageUrl:
            'https://ichef.test.bbci.co.uk/news/{width}/cpsdevpb/17654/test/_63482859_orange1.jpg',
          imageAlt: 'Orange 1',
          id: 'ae7c9f51-3144-ba4b-a881-f88822b36c8c',
        },
      ],
      activePage: 1,
      pageCount: 1,
      curationId: 'urn:bbc:vivo:curation:bf45c519-cfee-4627-812f-67800cdeb30c',
      curationType: 'vivo-stream',
      visualProminence: 'NORMAL',
    },
  ],
  metadata: { type: 'Topic' },
};

export const amharicOnlyTitle = {
  title: 'ዶናልድ ትራምፕ',
  curations: [
    {
      summaries: [
        {
          title: 'በጥቅም ላይ የዋሉ የጥንታዊ ግብፅ የህክምና ጥበባት',
          type: 'article',
          firstPublished: '2017-08-22T08:41:04.000Z',
          link: 'https://www.bbc.com/amharic/news-23143874',
          imageUrl: '',
          imageAlt: '',
          id: 'c31dd8c5-2e55-9143-8ed2-5f9c15d6c91c',
        },
      ],
      activePage: 1,
      pageCount: 1,
      curationId: 'urn:bbc:vivo:curation:bf45c519-cfee-4627-812f-54329cdeb30c',
      curationType: 'vivo-stream',
      visualProminence: 'NORMAL',
    },
  ],
  metadata: { type: 'Topic' },
};
