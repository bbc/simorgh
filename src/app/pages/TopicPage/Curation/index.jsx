import React from 'react';
import { arrayOf, oneOf, shape, string, number } from 'prop-types';
import pathOr from 'ramda/src/pathOr';
import VisuallyHiddenText from '#psammead/psammead-visually-hidden-text/src';
import CurationGrid from './CurationGrid';
import HierarchicalGrid from '../HierarchicalGrid';
import Subheading from './Subhead';

export const VISUAL_STYLE = {
  NONE: 'NONE',
  LINKS: 'LINKS',
  BANNER: 'BANNER',
};

export const VISUAL_PROMINANCE = {
  NORMAL: 'NORMAL',
  HIGH: 'HIGH',
};

// Maps a visual style and prominance to a component that renders that curation
const components = {
  [VISUAL_STYLE.NONE]: {
    [VISUAL_PROMINANCE.NORMAL]: CurationGrid,
    [VISUAL_PROMINANCE.HIGH]: HierarchicalGrid,
  },
};

const Curation = ({
  visualStyle = VISUAL_STYLE.NONE,
  visualProminance,
  promos,
  title,
  topStoriesTitle,
  link,
  headingLevel,
  position,
  curationLength,
}) => {
  if (visualStyle === VISUAL_STYLE.BANNER) return null;
  if (!promos.length) return null;
  const Component = pathOr(
    CurationGrid,
    [visualStyle, visualProminance],
    components,
  );

  const createID = titleText => {
    return titleText.replaceAll(' ', '-');
  };

  const isFirstCuration = position === 0;
  const SubheadingComponent = isFirstCuration ? VisuallyHiddenText : Subheading;
  const id = createID(title || topStoriesTitle);

  return curationLength > 1 && (title || isFirstCuration) ? (
    <section aria-labelledby={id} role="region">
      <SubheadingComponent as="h2" a11yID={id} id={id} link={link}>
        {title || topStoriesTitle}
      </SubheadingComponent>
      <Component
        promos={promos}
        headingLevel={isFirstCuration ? 3 : headingLevel}
      />
    </section>
  ) : (
    <Component promos={promos} headingLevel={headingLevel} />
  );
};

Curation.propTypes = {
  visualStyle: oneOf(Object.values(VISUAL_STYLE)).isRequired,
  visualProminance: oneOf(Object.values(VISUAL_PROMINANCE)).isRequired,
  promos: arrayOf(shape({})).isRequired,
  title: string,
  link: string,
  headingLevel: number,
  position: number.isRequired,
  topStoriesTitle: string,
  curationLength: number.isRequired,
};

Curation.defaultProps = {
  title: '',
  link: '',
  headingLevel: 2,
  topStoriesTitle: '',
};

export default Curation;
