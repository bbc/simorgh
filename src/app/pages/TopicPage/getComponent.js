import { VISUAL_PROMINENCE, VISUAL_STYLE, COMPONENT_NAMES } from './constants';

const { NONE, BANNER } = VISUAL_STYLE;
const { MINIMUM, LOW, NORMAL, HIGH, MAXIMUM } = VISUAL_PROMINENCE;
const {
  MESSAGE_BANNER,
  SIMPLE_CURATION_GRID,
  HIERARCHICAL_CURATION_GRID,
  NOT_SUPPORTED,
} = COMPONENT_NAMES;

export default (visualStyle, visualProminence) => {
  const componentsByVisualStyleAndProminence = {
    [`${BANNER}_${MINIMUM}`]: NOT_SUPPORTED,
    [`${BANNER}_${LOW}`]: NOT_SUPPORTED,
    [`${BANNER}_${NORMAL}`]: MESSAGE_BANNER,
    [`${BANNER}_${HIGH}`]: NOT_SUPPORTED,
    [`${BANNER}_${MAXIMUM}`]: NOT_SUPPORTED,
    [`${NONE}_${NORMAL}`]: SIMPLE_CURATION_GRID,
    [`${NONE}_${HIGH}`]: HIERARCHICAL_CURATION_GRID,
  };

  const visualStyleAndProminence = `${visualStyle}_${visualProminence}`;

  return componentsByVisualStyleAndProminence[visualStyleAndProminence] || null;
};
