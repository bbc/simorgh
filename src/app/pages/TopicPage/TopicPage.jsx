import React, { useContext } from 'react';
import ATIAnalytics from '#containers/ATIAnalytics';
import { shape, arrayOf, string } from 'prop-types';
import path from 'ramda/src/path';
import AdContainer from '#containers/Ad';
import CanonicalAdBootstrapJs from '#containers/Ad/Canonical/CanonicalAdBootstrapJs';
import useToggle from '#hooks/useToggle';
import { RequestContext } from '#contexts/RequestContext';
import ChartbeatAnalytics from '#containers/ChartbeatAnalytics';
import Curation from '#app/components/Curation';
import LinkedData from '../../components/LinkedData';
import styles from './index.styles';
import MetadataContainer from '../../components/Metadata';
import { ServiceContext } from '../../contexts/ServiceContext';
import TopicImage from './TopicImage';
import TopicTitle from './TopicTitle';
import TopicDescription from './TopicDescription';
import Pagination from './Pagination';

const TopicPage = ({ pageData }) => {
  const { lang, translations } = useContext(ServiceContext);
  const { title, description, imageData, curations, pageCount, activePage } =
    pageData;

  const { enabled: adsEnabled } = useToggle('ads');
  const { showAdsBasedOnLocation } = useContext(RequestContext);
  const topStoriesTitle = path(['topStoriesTitle'], translations);

  const linkedDataEntities = curations
    .map(({ summaries }) =>
      summaries.map(summary => ({
        '@type': summary.type,
        name: summary.title,
        headline: summary.title,
        url: summary.link,
        dateCreated: summary.firstPublished,
      })),
    )
    .flat();

  const { pageXOfY, previousPage, nextPage, page } = {
    pageXOfY: 'Page {x} of {y}',
    previousPage: 'Previous Page',
    nextPage: 'Next Page',
    page: 'Page',
    ...translations.pagination,
  };

  const translatedPage = pageXOfY
    .replace('{x}', activePage)
    .replace('{y}', pageCount);

  const pageTitle = `${title}, ${translatedPage}`;

  return (
    <>
      {adsEnabled && showAdsBasedOnLocation && (
        <>
          <CanonicalAdBootstrapJs />
          <AdContainer slotType="leaderboard" />
        </>
      )}
      <main css={styles.main}>
        <div css={styles.inner}>
          <ATIAnalytics data={pageData} />
          <ChartbeatAnalytics data={pageData} />
          <MetadataContainer
            title={activePage >= 2 ? pageTitle : title}
            socialHeadline={title}
            lang={lang}
            description={description}
            openGraphType="website"
            hasAmpPage={false}
          />
          <LinkedData
            type="CollectionPage"
            seoTitle={title}
            headline={title}
            entities={linkedDataEntities}
          />
          <div css={styles.title}>
            <div css={styles.inline}>
              {imageData && <TopicImage image={imageData.url} />}
              <TopicTitle>{title}</TopicTitle>
            </div>
            {description && <TopicDescription>{description}</TopicDescription>}
          </div>
          {curations.map(
            ({
              visualProminence,
              summaries,
              curationId,
              title: curationTitle,
              link,
              position,
              visualStyle,
            }) => {
              return (
                <React.Fragment key={`${curationId}-${position}`}>
                  <Curation
                    headingLevel={curationTitle && 3}
                    visualStyle={visualStyle}
                    visualProminence={visualProminence}
                    promos={summaries}
                    title={curationTitle}
                    topStoriesTitle={topStoriesTitle}
                    position={position}
                    link={link}
                    curationLength={curations && curations.length}
                  />
                </React.Fragment>
              );
            },
          )}
          <Pagination
            activePage={activePage}
            pageCount={pageCount}
            pageXOfY={pageXOfY}
            previousPage={previousPage}
            nextPage={nextPage}
            page={page}
          />
        </div>
      </main>
    </>
  );
};

TopicPage.propTypes = {
  pageData: shape({
    title: string.isRequired,
    curations: arrayOf(shape({})).isRequired,
  }).isRequired,
};

export default TopicPage;
