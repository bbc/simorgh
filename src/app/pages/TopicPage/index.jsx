import TopicPage from './TopicPage';
import applyBasicPageHandlers from '../utils/applyBasicPageHandlers';

export default applyBasicPageHandlers(TopicPage);
