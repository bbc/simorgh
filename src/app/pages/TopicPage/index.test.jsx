import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { TOPIC_PAGE } from '#app/routes/utils/pageTypes';
import { data as kyrgyzTopicWithMessageBanners } from '#data/kyrgyz/topics/cvpv9djp9qqt.json';
import { data as mundoBannerVariations } from '#data/mundo/topics/cw90edn9kw4t.json';
import {
  VISUAL_PROMINENCE,
  VISUAL_STYLE,
} from '#app/models/types/curationData';
import { render } from '../../components/react-testing-library-with-providers';
import TopicPage from './TopicPage';
import {
  pidginMultipleItems,
  amharicSingleItem,
  mundoWithBadgeAndDescr,
  mundoMultipleCurations,
  amharicOnlyTitle,
} from './fixtures';

jest.mock('../../components/ThemeProvider');
jest.mock('../../legacy/containers/ChartbeatAnalytics', () => {
  const ChartbeatAnalytics = () => <div>chartbeat</div>;
  return ChartbeatAnalytics;
});

const getOptionParams = ({
  showAdsBasedOnLocation = false,
  service = 'pidgin',
  lang = 'pcm',
  adsToggledOn = false,
} = {}) => ({
  isAmp: false,
  showAdsBasedOnLocation,
  pageType: TOPIC_PAGE,
  service,
  lang,
  toggles: {
    ads: {
      enabled: adsToggledOn,
    },
  },
});

describe('Topic Page', () => {
  it('should not render an unordered list when there is only one promo', () => {
    const { queryByRole } = render(
      <TopicPage pageData={amharicSingleItem} />,
      getOptionParams({ service: 'amharic', lang: 'am' }),
    );
    expect(queryByRole('list')).not.toBeInTheDocument();
  });

  it('should render an unordered list when there is more than one promo', () => {
    const { container, queryByRole } = render(
      <TopicPage pageData={pidginMultipleItems} />,
      getOptionParams(),
    );
    expect(queryByRole('list')).toBeInTheDocument();
    expect(container.getElementsByTagName('li').length).toEqual(4);
  });

  it('should render multiple curations', () => {
    const { getAllByRole } = render(
      <TopicPage pageData={mundoMultipleCurations} />,
      getOptionParams({ service: 'mundo', lang: 'es' }),
    );
    expect(getAllByRole('list').length).toEqual(2);
    expect(getAllByRole('listitem').length).toEqual(4);
  });

  it('should render a section around each curation when more than one exists', () => {
    const { container } = render(
      <TopicPage pageData={mundoMultipleCurations} />,
      getOptionParams({ service: 'mundo', lang: 'es' }),
    );
    expect(container.getElementsByTagName('section').length).toEqual(2);
  });

  it('should not render a section when one or less exists', () => {
    const { container } = render(
      <TopicPage pageData={amharicSingleItem} />,
      getOptionParams({ service: 'amharic', lang: 'am' }),
    );
    expect(container.getElementsByTagName('section').length).toEqual(0);
  });

  it('should render curation subheading when curation title exists', () => {
    const { container } = render(
      <TopicPage pageData={mundoMultipleCurations} />,
      getOptionParams({ service: 'mundo', lang: 'es' }),
    );
    expect(container.querySelector('h2').textContent).toEqual('Analysis');
  });

  it('should render promo headings as h3 when curation subheading exists', () => {
    const { container } = render(
      <TopicPage pageData={mundoMultipleCurations} />,
      getOptionParams({ service: 'mundo', lang: 'es' }),
    );

    expect(container.getElementsByTagName('h3').length).toEqual(4);
    expect(container.getElementsByTagName('h2').length).toEqual(2);
  });

  it('should render promo headings as h2 when there is no curation subheading', () => {
    const { container } = render(
      <TopicPage pageData={pidginMultipleItems} service="pidgin" />,
      getOptionParams(),
    );

    expect(container.getElementsByTagName('h3').length).toEqual(0);
    expect(container.getElementsByTagName('h2').length).toEqual(4);
  });

  it('should render badge and description when they exist in data', () => {
    const { container, queryByTestId } = render(
      <TopicPage pageData={mundoWithBadgeAndDescr} />,
      getOptionParams({ service: 'mundo', lang: 'es' }),
    );

    expect(queryByTestId('topic-badge')).toBeInTheDocument();
    expect(container.getElementsByTagName('p').length).toEqual(1);
  });

  it('should render description without badge', () => {
    const { container, queryByTestId } = render(
      <TopicPage pageData={pidginMultipleItems} service="pidgin" />,
      getOptionParams(),
    );

    expect(queryByTestId('topic-badge')).not.toBeInTheDocument();
    expect(container.getElementsByTagName('p').length).toEqual(1);
  });

  it('should render only topic title without badge or description', () => {
    const { container, queryByTestId } = render(
      <TopicPage pageData={amharicOnlyTitle} />,
      getOptionParams({ service: 'amharic', lang: 'am' }),
    );

    expect(queryByTestId('topic-badge')).not.toBeInTheDocument();
    expect(container.getElementsByTagName('p').length).toEqual(0);
  });

  it('should show ads when enabled', () => {
    [
      [true, true],
      [true, false],
      [false, true],
      [false, false],
    ].forEach(([adsToggledOn, showAdsBasedOnLocation]) => {
      const { container } = render(
        <BrowserRouter>
          <TopicPage pageData={pidginMultipleItems} />
        </BrowserRouter>,
        getOptionParams({ adsToggledOn, showAdsBasedOnLocation }),
      );

      const shouldShowAds = adsToggledOn && showAdsBasedOnLocation;
      const adElement = container.querySelector('[data-e2e="advertisement"]');
      if (shouldShowAds) {
        expect(adElement).toBeInTheDocument();
      } else {
        expect(adElement).not.toBeInTheDocument();
      }
    });
  });

  describe('Message Banner', () => {
    it('should only render when visual style is banner and visual prominence is normal', () => {
      const messageBannerCuration =
        kyrgyzTopicWithMessageBanners.curations.find(
          ({ visualStyle, visualProminence }) =>
            visualProminence === VISUAL_PROMINENCE.NORMAL &&
            visualStyle === VISUAL_STYLE.BANNER,
        );

      const { getByRole } = render(
        <TopicPage pageData={kyrgyzTopicWithMessageBanners} />,
        getOptionParams({ service: 'kyrgyz', lang: 'ky' }),
      );

      const messageBanner = getByRole('region', {
        name: messageBannerCuration.title,
      });

      expect(messageBanner).toBeInTheDocument();
    });

    it('should render multiple message banners with unique ids', () => {
      render(
        <TopicPage pageData={mundoBannerVariations} />,
        getOptionParams({ service: 'mundo', lang: 'es' }),
      );
      expect(document.querySelectorAll("[id^='message-banner']")).toHaveLength(
        3,
      );
      expect(document.querySelectorAll("[id='message-banner']")).toHaveLength(
        0,
      );
    });

    it('should not render when visual style is banner and visual prominence is high', () => {
      const { queryByRole } = render(
        <TopicPage pageData={kyrgyzTopicWithMessageBanners} />,
        getOptionParams({ service: 'kyrgyz', lang: 'ky' }),
      );
      const highProminenceBanner = kyrgyzTopicWithMessageBanners.curations.find(
        curation =>
          curation.visualStyle === VISUAL_STYLE.BANNER &&
          curation.visualProminence === VISUAL_PROMINENCE.HIGH,
      );
      expect(
        queryByRole('region', { name: highProminenceBanner.title }),
      ).not.toBeInTheDocument();
    });

    it('should not render when visual style is not banner', () => {
      render(
        <TopicPage pageData={amharicSingleItem} />,
        getOptionParams({ service: 'amharic', lang: 'am' }),
      );

      expect(document.querySelectorAll("[id^='message-banner']")).toHaveLength(
        0,
      );
    });

    it('should only render the first summary if there is more than one summary in the curation', () => {
      const messageBannerCuration =
        kyrgyzTopicWithMessageBanners.curations.find(
          ({ visualStyle, visualProminence, summaries }) =>
            visualProminence === VISUAL_PROMINENCE.NORMAL &&
            visualStyle === VISUAL_STYLE.BANNER &&
            summaries.length > 1,
        );
      const { queryAllByRole } = render(
        <TopicPage pageData={kyrgyzTopicWithMessageBanners} />,
        getOptionParams({ service: 'kyrgyz', lang: 'ky' }),
      );
      const messageBanners = queryAllByRole('region', {
        name: messageBannerCuration.title,
      });
      expect(messageBanners).toHaveLength(1);
    });
  });
});
