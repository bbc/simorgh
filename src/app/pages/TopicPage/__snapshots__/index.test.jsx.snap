// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topic Page SEO should correctly render linked data 1`] = `
{
  "@context": "http://schema.org",
  "@graph": [
    {
      "@type": "CollectionPage",
      "headline": "Donald Trump",
      "image": {
        "@type": "ImageObject",
        "height": 576,
        "url": "https://news.files.bbci.co.uk/ws/img/logos/og/pidgin.png",
        "width": 1024,
      },
      "inLanguage": {
        "@type": "Language",
        "alternateName": "pcm",
        "name": "Nigerian Pidgin",
      },
      "mainEntityOfPage": {
        "@id": "https://www.test.bbc.com/news/articles/c0g992jmmkko",
        "@type": "WebPage",
        "name": "Donald Trump",
      },
      "publisher": {
        "@type": "NewsMediaOrganization",
        "logo": {
          "@type": "ImageObject",
          "height": 576,
          "url": "https://news.files.bbci.co.uk/ws/img/logos/og/pidgin.png",
          "width": 1024,
        },
        "name": "BBC News Pidgin",
        "publishingPrinciples": "https://www.bbc.com/pidgin/institutional-48528766",
      },
      "thumbnailUrl": "https://news.files.bbci.co.uk/ws/img/logos/og/pidgin.png",
      "url": "https://www.test.bbc.com/news/articles/c0g992jmmkko",
    },
    {
      "@type": "ItemList",
      "itemListElement": [
        {
          "@context": "http://schema.org",
          "@type": "ListItem",
          "position": 1,
          "url": "https://www.bbc.com/pidgin/world-60465144",
        },
        {
          "@context": "http://schema.org",
          "@type": "ListItem",
          "position": 2,
          "url": "https://www.bbc.com/pidgin/tori-59901959",
        },
        {
          "@context": "http://schema.org",
          "@type": "ListItem",
          "position": 3,
          "url": "https://www.bbc.com/pidgin/tori-57396574",
        },
        {
          "@context": "http://schema.org",
          "@type": "ListItem",
          "position": 4,
          "url": "https://www.bbc.com/pidgin/tori-56991816",
        },
      ],
      "name": "BBC News Pidgin",
      "numberOfItems": 4,
    },
  ],
}
`;
